File: Machine_Learning_Algorithms/Java/k_means_clustering.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

public class k_means_clustering {

    static class KMeans {
        int k;
        int maxIterations;
        double[][] centroids;
        int[] labels;
        Random random;

        public KMeans(int k, int maxIterations) {
            this.k = k;
            this.maxIterations = maxIterations;
            this.random = new Random();
        }

        private void initializeCentroids(double[][] X) {
            // Chọn ngẫu nhiên k điểm dữ liệu làm tâm cụm ban đầu
            int nSamples = X.length;
            centroids = new double[k][X[0].length];
            List<Integer> chosenIndices = new ArrayList<>();

            for (int i = 0; i < k; i++) {
                int randomIndex;
                do {
                    randomIndex = random.nextInt(nSamples);
                } while (chosenIndices.contains(randomIndex));
                chosenIndices.add(randomIndex);
                centroids[i] = Arrays.copyOf(X[randomIndex], X[randomIndex].length);
            }
        }

        private void assignClusters(double[][] X) {
            // Gán mỗi điểm dữ liệu cho tâm cụm gần nhất
            labels = new int[X.length];
            for (int i = 0; i < X.length; i++) {
                double minDistance = Double.MAX_VALUE;
                int closestCentroidIndex = -1;
                for (int j = 0; j < k; j++) {
                    double distance = euclideanDistance(X[i], centroids[j]);
                    if (distance < minDistance) {
                        minDistance = distance;
                        closestCentroidIndex = j;
                    }
                }
                labels[i] = closestCentroidIndex;
            }
        }

        private void updateCentroids(double[][] X) {
            // Cập nhật tâm cụm bằng cách tính trung bình các điểm trong mỗi cụm
            double[][] newCentroids = new double[k][X[0].length];
            int[] clusterCounts = new int[k];

            for (int i = 0; i < X.length; i++) {
                int clusterIndex = labels[i];
                for (int j = 0; j < X[0].length; j++) {
                    newCentroids[clusterIndex][j] += X[i][j];
                }
                clusterCounts[clusterIndex]++;
            }

            for (int i = 0; i < k; i++) {
                if (clusterCounts[i] > 0) {
                    for (int j = 0; j < X[0].length; j++) {
                        newCentroids[i][j] /= clusterCounts[i];
                    }
                } else {
                    // Nếu một cụm trống, khởi tạo lại tâm cụm ngẫu nhiên
                    int randomIndex = random.nextInt(X.length);
                    newCentroids[i] = Arrays.copyOf(X[randomIndex], X[randomIndex].length);
                }
            }
            centroids = newCentroids;
        }

        private double euclideanDistance(double[] p1, double[] p2) {
            double sum = 0;
            for (int i = 0; i < p1.length; i++) {
                sum += Math.pow(p1[i] - p2[i], 2);
            }
            return Math.sqrt(sum);
        }

        public void fit(double[][] X) {
            initializeCentroids(X);

            for (int iter = 0; iter < maxIterations; iter++) {
                double[][] oldCentroids = Arrays.stream(centroids).map(double[]::clone).toArray(double[][]::new);
                assignClusters(X);
                updateCentroids(X);

                // Kiểm tra sự hội tụ
                boolean converged = true;
                for (int i = 0; i < k; i++) {
                    if (euclideanDistance(oldCentroids[i], centroids[i]) > 1e-6) { // Sử dụng ngưỡng nhỏ để so sánh số thực
                        converged = false;
                        break;
                    }
                }
                if (converged) {
                    break;
                }
            }
        }

        public int[] predict(double[][] X) {
            assignClusters(X);
            return labels;
        }
    }

    public static void main(String[] args) {
        // Ví dụ sử dụng:
        // Tạo dữ liệu giả
        double[][] X = {
            {1, 2}, {1.5, 1.8}, {5, 8}, {8, 8}, {1, 0.6}, {9, 11}
        };

        KMeans kmeans = new KMeans(2, 100);
        kmeans.fit(X);

        System.out.println("Tâm cụm cuối cùng:");
        for (double[] centroid : kmeans.centroids) {
            System.out.println(Arrays.toString(centroid));
        }
        System.out.println("Nhãn cụm cho mỗi điểm dữ liệu: " + Arrays.toString(kmeans.labels));
    }
}

File: Machine_Learning_Algorithms/JavaScript/k_means_clustering.js
class KMeans {
    constructor(k = 3, maxIterations = 100) {
        this.k = k;
        this.maxIterations = maxIterations;
        this.centroids = null;
        this.labels = null;
    }

    _initializeCentroids(X) {
        // Chọn ngẫu nhiên k điểm dữ liệu làm tâm cụm ban đầu
        const nSamples = X.length;
        this.centroids = new Array(this.k).fill(0).map(() => new Array(X[0].length).fill(0));
        const chosenIndices = new Set();

        for (let i = 0; i < this.k; i++) {
            let randomIndex;
            do {
                randomIndex = Math.floor(Math.random() * nSamples);
            } while (chosenIndices.has(randomIndex));
            chosenIndices.add(randomIndex);
            this.centroids[i] = [...X[randomIndex]];
        }
    }

    _assignClusters(X) {
        // Gán mỗi điểm dữ liệu cho tâm cụm gần nhất
        this.labels = new Array(X.length);
        for (let i = 0; i < X.length; i++) {
            let minDistance = Infinity;
            let closestCentroidIndex = -1;
            for (let j = 0; j < this.k; j++) {
                const distance = this._euclideanDistance(X[i], this.centroids[j]);
                if (distance < minDistance) {
                    minDistance = distance;
                    closestCentroidIndex = j;
                }
            }
            this.labels[i] = closestCentroidIndex;
        }
    }

    _updateCentroids(X) {
        // Cập nhật tâm cụm bằng cách tính trung bình các điểm trong mỗi cụm
        const newCentroids = new Array(this.k).fill(0).map(() => new Array(X[0].length).fill(0));
        const clusterCounts = new Array(this.k).fill(0);

        for (let i = 0; i < X.length; i++) {
            const clusterIndex = this.labels[i];
            for (let j = 0; j < X[0].length; j++) {
                newCentroids[clusterIndex][j] += X[i][j];
            }
            clusterCounts[clusterIndex]++;
        }

        for (let i = 0; i < this.k; i++) {
            if (clusterCounts[i] > 0) {
                for (let j = 0; j < X[0].length; j++) {
                    newCentroids[i][j] /= clusterCounts[i];
                }
            } else {
                // Nếu một cụm trống, khởi tạo lại tâm cụm ngẫu nhiên
                const randomIndex = Math.floor(Math.random() * X.length);
                newCentroids[i] = [...X[randomIndex]];
            }
        }
        this.centroids = newCentroids;
    }

    _euclideanDistance(p1, p2) {
        let sum = 0;
        for (let i = 0; i < p1.length; i++) {
            sum += Math.pow(p1[i] - p2[i], 2);
        }
        return Math.sqrt(sum);
    }

    fit(X) {
        this._initializeCentroids(X);

        for (let iter = 0; iter < this.maxIterations; iter++) {
            const oldCentroids = JSON.parse(JSON.stringify(this.centroids)); // Deep copy
            this._assignClusters(X);
            this._updateCentroids(X);

            // Kiểm tra sự hội tụ
            let converged = true;
            for (let i = 0; i < this.k; i++) {
                if (this._euclideanDistance(oldCentroids[i], this.centroids[i]) > 1e-6) {
                    converged = false;
                    break;
                }
            }
            if (converged) {
                break;
            }
        }
    }

    predict(X) {
        this._assignClusters(X);
        return this.labels;
    }
}

// Ví dụ sử dụng:
// Tạo dữ liệu giả
const X = [
    [1, 2],
    [1.5, 1.8],
    [5, 8],
    [8, 8],
    [1, 0.6],
    [9, 11]
];

const kmeans = new KMeans(2, 100);
kmeans.fit(X);

console.log("Tâm cụm cuối cùng:", kmeans.centroids);
console.log("Nhãn cụm cho mỗi điểm dữ liệu:", kmeans.labels);
